proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+- Inline property delegates
+```kotlin
+class A {
+    var something by InlinedDelegate(Foo()) // no actual instantiation of `InlinedDelegate`
+}
+
+
+inline class InlinedDelegate<T>(var node: T) {
+    operator fun setValue(thisRef: A, property: KProperty<*>, value: T) {
+        if (node !== value) {
+            thisRef.notify(node, value)
+        }
+        node = value
+    }
+
+    operator fun getValue(thisRef: A, property: KProperty<*>): T {
+        return node
+    }
+}
+```
+
+- Inline wrappers
+    - Typed wrappers
+    ```kotlin
+    inline class Name(private val s: String)
+    inline class Password(private val s: String)
+    
+    fun foo() {
+        var n = Name("n") // no actual instantiation, on JVM type of `n` is String
+        val p = Password("p")
+        n = "other" // type mismatch error
+        n = p // type mismatch error
+    }
+    ```
+
+    - API refinement
+    ```
+    // Java
+    public class Foo {
+        public Object[] objects() { ... }
+    }
+    
+    // Kotlin
+    inline class RefinedFoo(val f: Foo) {
+        inline fun <T> array(): Array<T> = f.objects() as Array<T>
+    }
+    ``` 
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```kotlin
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```kotlin
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
+- Underlying value cannot be of the same type that is containing inline class
+- Inline class with undefined (recursively defined) generics, e.g. generics with an upper bound equal to the class, is prohibited
+    ```kotlin
+    inline class A<T : A<T>>(val x: T) // error
+    ```
+- Inline class cannot have `init` block
+- Inline class must be final
+- Inline class can implement only interfaces
+- Inline class cannot have backing fields
+    - Hence, it follows that inline class can have only simple computable properties (no lateinit/delegated properties)
+- Inline class cannot have inner classes
+- Inline class must be a toplevel class 
```

Comment:
hm, that's pretty limiting, why is that?

Use-case: inside a class I want to use an inline class as an implementation detail, but I don't want other classes to know about it

Reply:
Indeed, this is too limiting, thanks! 
It's definitely can be a nested member, I'll think more about it